compiler/z/codegen/BinaryEvaluator.cpp

```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
             {
             traceMsg(comp,"[%p] land => rotated-and-insert: tZeros %d, lZeros %d, popCnt %d\n", node, tZeros, lZeros, popCnt);
             traceMsg(comp,"          => rotated-and-insert: lOnes %d, tOnes %d\n", lOnes, tOnes);
             }
 
-         bool doIt = false;
+         bool doTransformation = false;
+
+         // We now try to detect if the lconst's bit pattern is such that we can use a single RISBG
+         // instead of multiple AND instructions
+
+         // The first block will catch all cases where the lConstValue is a series of contiguous ones 
+         // surrounded by zeros
+         // ex: 0001111000, 011110, 111111, 001111, 11100 etc...
+         // 
+         // The 'else if' block will consider all cases where we have a series of contiguous zeros 
+         // surrounded by ones
+         // ex: 1110000111, 1000001
+         //
+         // Note: if the lconst is 0x0 we cannot generate a RISBG instruction.
+         // The starting and ending bit positions are inclusive, so specifying (lsBit,msBit)=(0,0) will
+         // preserve the 0th bit in the firstChild, which would be incorrect.
+         // If such a case arises where we can figure out at compile time that the 
+         // constant's value is 0, then we must handle that case in the simplifier.
          if (popCnt == (64 - lZeros - tZeros))
             {
-            msBit = lZeros;
+            msBit = lZeros; 
             lsBit = 63 - tZeros;
-            if (firstChild->getReferenceCount()>1 || skipedConversion != NULL   // Better of with non-destructive RISBG as AND?
-                  || lZeros > 31 || tZeros > 31 || (lZeros > 0 && tZeros > 0))  // Better off with a single AND?
+            if (firstChild->getReferenceCount() > 1 // if we use an AND and refCount>1, then we will have to backup the value in another register beforehand, as AND is destructive
+                || lZeros > 31 || tZeros > 31 // this makes sure we are not better off using one of the following AND instructions: NILF,NILH,NILL, NIHF,NIHH, NIHL
+                || (lZeros > 0 && tZeros > 0)) // if our constant at minimum looks like 0111...110, then it's better to use RISBG as this case will require multiple ANDs
                {
-               doIt = true;
+               doTransformation = true;
                }
-            }
-         else if (popCnt == (lOnes + tOnes))
+            }                         
+         else if (popCnt == (lOnes + tOnes) && popCnt > 0) 
             {
             msBit = 64 - tOnes;
             lsBit = lOnes - 1;
-            if (firstChild->getReferenceCount()>1 || skipedConversion != NULL // Better of with non-destructive RISBG as AND?
-                  || (lOnes < 32 && tOnes < 32) )                             // Better off with a single AND?
+            if (firstChild->getReferenceCount()>1 // if we use an AND and refCount>1, then we will have to backup the value in another register beforehand, as AND is destructive
```

Comment:
Same as above. Can we not use clobber evaluate if it will save generating two ANDs?

Reply:
As mentioned in another comment, I don't think we need clobber evaluate. We don't need to store the original value in another register because the RISBG instruction is not destructive. So we can just store the result of the RISBG instruction in another register. I would think a regular evaluate is probably faster in terms of compile time performance as well?

Reply:
> I would think a regular evaluate is probably faster in terms of compile time performance as well?

Yes we just need a regular evaluation of the first child. I don't think we need to reference count check here as a guard.